### PR TITLE
Enable acked indexing

### LINKED ISSUE #7572: [Indexing] A network partition can cause in flight documents to be lost
This ticket is meant to capture an issue which was discovered as part of the work done in #7493 , which contains a [failing reproduction test](https://github.com/elasticsearch/elasticsearch/blob/596a4a073584c4262d574828c9caea35b5ed1de5/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptions.java#L375) with @awaitFix.

If a network partition separates a node from the master, there is some window of time before the node detects it. The length of the window is dependent on the type of the partition. This window is extremely small if a socket is broken. More adversarial partitions, for example, silently dropping requests without breaking the socket can take longer (up to 3x30s using current defaults).

If the node hosts a *primary* shard at the moment of partition, and ends up being isolated from the cluster (which could have resulted in Split Brain before), some documents that are being indexed into the primary *may* be lost if they fail to reach one of the allocated replicas  (due to the partition) and that replica is later promoted to primary by the master.

### PR DESCRIPTION
**Pull request by jasontedor**
Recent work on the distributed systems front brings us to a point where
we are very close to being able to enable the acked indexing test. We
have chased down one additional failure that will require the primary
terms work to address, but let's get this preparatory work in.

Closes #7572

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 1]
```diff
@@ -820,24 +821,64 @@ void finishBecauseUnavailable(ShardId shardId, String message) {
         }
     }
 
+
+    static ConcurrentMap<IndexShardReference, String> openShardReferences;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: This feels way more natural on the IndexShard level - I felt that way  when I made it but couldn't find a way (at the time) to push it to IndexShard while getting the debugging information to be used as the string value here. We can pass it to IndexShard.acquirePrimaryOperationLock and acquireReplicaOperationLock but that will force a toString on the request object. How about making the reason an Object type and only call toString on it in IndexShard?

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: or even better a reason supplier called on demand?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 2]
```diff
@@ -1018,30 +1059,36 @@ public void handleException(TransportException exp) {
...
+
+                                            @Override
+                                            public void onFailure(Throwable shardFailedError) {
+                                                if (shardFailedError instanceof ShardStateAction.NoLongerPrimaryShardException) {
+                                                    String message = "unknown";
+                                                    try {
+                                                        ShardRouting primaryShard = indexShardReference.routingEntry();
+                                                        message = String.format(Locale.ROOT, "primary shard [%s] was demoted while failing replica shard [%s] for [%s]", primaryShard, shard, exp);
+                                                        // we are no longer the primary, fail ourselves and start over
+                                                        indexShardReference.failShard(message, shardFailedError);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: can we open a follow up issue that  failShard should throw an already closed exception? (this is really what the code protects against). To be clear - I think the code can stay.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 3]
```diff
@@ -1018,30 +1059,36 @@ public void handleException(TransportException exp) {
...
+                                                    } catch (Throwable t) {
+                                                        shardFailedError.addSuppressed(t);
+                                                    }
+                                                    forceFinishAsFailed(new RetryOnPrimaryException(shardId, message, shardFailedError));
+                                                } else {
+                                                    assert shardFailedError instanceof TransportException ||
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: add a comment about where these exceptions can come from?

File: core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java [hunk 4]
```diff
@@ -223,9 +223,10 @@ protected void doStop() {
         FutureUtils.cancel(this.reconnectToNodes);
         for (NotifyTimeout onGoingTimeout : onGoingTimeouts) {
             onGoingTimeout.cancel();
-            onGoingTimeout.listener.onClose();
         }
         ThreadPool.terminate(updateTasksExecutor, 10, TimeUnit.SECONDS);
+        postAppliedListeners.stream().filter(listener -> listener instanceof TimeoutClusterStateListener)
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java]: add a comment why we do it out of loop? (i.e. loop again)

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java [hunk 5]
```diff
@@ -402,6 +403,18 @@ void validateIncomingState(ClusterState incomingState, ClusterState lastSeenClus
         }
 
         ZenDiscovery.validateStateIsFromCurrentMaster(logger, currentNodes, incomingState);
+        if (lastSeenClusterState != null && lastSeenClusterState.supersedes(incomingState)) {
+            final String message = String.format(
+                    Locale.ROOT,
+                    "received older cluster state version [%s] with uuid [%s] than last seen cluster state [%s] with uuid [%s]",
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: can we add that it came from the current master? Also - thinking about this more - this can only happen in our testing right? In practice we only have 1 channel (if someone doesn't mess with some settings). Maybe add a comment about it if so...

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 6]
```diff
@@ -759,7 +759,13 @@ public void releaseSnapshot(IndexCommit snapshot) throws IOException {
      */
     public void failShard(String reason, @Nullable Throwable e) {
         // fail the engine. This will cause this shard to also be removed from the node's index service.
-        getEngine().failEngine(reason, e);
+        final Engine engine = getEngineOrNull();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: haha. I fixed this inline - forgot :) I think this should be a separate PR - it's not really needed for this one (we protect for it).

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java [hunk 7]
```diff
@@ -139,6 +138,17 @@ protected int numberOfReplicas() {
         return 1;
     }
 
+    @Override
+    protected void beforeIndexDeletion() {
+        // some test may leave operations in flight
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java]: I think it will be good to explain here that unresponsive disruption swallows request by design, meaning that some operations will never be marked as finished.

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java [hunk 8]
```diff
@@ -453,36 +467,33 @@ public void testAckedIndexing() throws Exception {
...
-                                    logger.trace("[{}] decreased counter : {}", name, countDownLatchRef.get().getCount());
-                                }
-                            } catch (InterruptedException e) {
-                                // fine - semaphore interrupt
-                            } catch (Throwable t) {
-                                logger.info("unexpected exception in background thread of [{}]", t, node);
+                                id = Integer.toString(idGenerator.incrementAndGet());
+                                int shard = MathUtils.mod(Murmur3HashFunction.hash(id), numPrimaries);
+                                logger.trace("[{}] indexing id [{}] through node [{}] targeting shard [{}]", name, id, node, shard);
+                                IndexResponse response = client.prepareIndex("test", "type", id).setSource("{}").setTimeout("1s").get("1s");
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java]: I wonder if should rarely (and if nightly runs are enable) run with a longer timeout, say 5s...

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java [hunk 9]
```diff
@@ -453,36 +467,33 @@ public void testAckedIndexing() throws Exception {
...
+                                assertThat(response.getVersion(), equalTo(1L));
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java]: nit: I think this is cleaner

```
                                    assertTrue("doc ["+id+"] should have been created", response.isCreated())
```

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java [hunk 10]
```diff
@@ -512,11 +523,15 @@ public void run() {
                     assertThat(semaphore.availablePermits(), equalTo(0));
                     semaphore.release(docsPerIndexer);
                 }
-                assertTrue(countDownLatchRef.get().await(60000 + disruptionScheme.expectedTimeToHeal().millis() * (docsPerIndexer * indexers.size()), TimeUnit.MILLISECONDS));
+                logger.info("waiting for indexing requests to complete");
+                assertTrue(countDownLatchRef.get().await(docsPerIndexer * 1000 + 2000, TimeUnit.MILLISECONDS));
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java]: hmm... I'm not sure about this. It means we are not indexing when the partition is healed... not sure why I changed it. I think it was to make sure that we have some indexing during disruption but looking at it now I think we should split into rounds - indexing while disrupting , stop disrupt and index some more, then wait for the cluster to heal. WDYT?

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java [hunk 11]
```diff
@@ -843,17 +867,21 @@ public void testClusterJoinDespiteOfPublishingIssues() throws Exception {
...
         nonMasterTransportService.addFailToSendNoConnectRule(masterTranspotService);
 
         assertNoMaster(nonMasterNode);
 
         logger.info("blocking cluster state publishing from master [{}] to non master [{}]", masterNode, nonMasterNode);
-        MockTransportService masterTransportService = (MockTransportService) internalCluster().getInstance(TransportService.class, masterNode);
-        TransportService localTransportService = internalCluster().getInstance(TransportService.class, discoveryNodes.localNode().getName());
+        MockTransportService masterTransportService = (MockTransportService) internalCluster().getInstance(TransportService.class,
+                masterNode);
+        TransportService localTransportService = internalCluster().getInstance(TransportService.class, discoveryNodes.localNode().getName
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java]: I hate IntelliJ's auto trimming. I have no idea who came up with that logic..

File: core/src/test/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateActionTests.java [hunk 12]
```diff
@@ -658,8 +670,17 @@ public void testInterleavedPublishCommit() throws Throwable {
         final ClusterState finalState = states.get(numOfStates - 1);
         Collections.shuffle(states, random());
 
-        logger.info("--> publishing states");
+        List<ClusterState> orderedSubsequence = new ArrayList<>();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateActionTests.java]: why do we need the subsequence? the original list is already ordered - we can just use it? also note that strictly speaking we do support out of order publishing (until we remove the settings that allows it, one day.. :)).

**Comment by bleskes** [comment]
Left a bunch of minor comments. Thanks for picking this up!

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java (hunk 8 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java]: I pushed 14ba0c31b4a90595efdfa64f3de23904a354f1dc.

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java (hunk 9 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java]: I pushed 37d739a3cdde49e10dc30c6c94eb90447656fd42.

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java (hunk 11 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java]: Down with unrelated formatting changes! (I pushed 97be38353a936ebe245f4fe7f1d311c82f784b11.)

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 3 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: Pushed 4e1f62eae93a8e1d447c1e7884ef2fb8fa037476.

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java (hunk 7 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java]: I pushed 2a9388912bc7e73381eb8b4d39fe03462abc7cb9.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 1 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: I agree that it's more natural there, but I prefer to keep it where it is until there's a clear need for it at the lower level.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 6 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I opened #17366.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 2 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: I opened #17366.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 6 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Reverted in 5576526c9182877184b2d347ca0eece875e0804f.

File: core/src/test/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateActionTests.java (hunk 12 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateActionTests.java]: I pushed 4de57fc5aac31c3a907529a1b932c934cdf611f6.

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java (hunk 5 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: Pushed 7f760347abbcb6eef46493c98b2171cd6f4d6748.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: guys we can't have static maps on classes like this. If we want some kind of static assertions we should inject some mock services just as we do for for `SearchContext` in `MockSearchService`.  I am a bit confused why we have a static factory method on literally the only impl of an interface in  `IndexShardReferenceImpl` this smells like a design flaw to me. I think with this change together we should rather expose a pluggalbe service or move the creation into `IndexService` and let folks plug in a factory into `IndexModule` in order to impl this assertion? but static per JVM concurrent maps seems like a playing russian roulette

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: Removed in 37d6c8c2e4016022526ee0d6ba2b0adaac23a7b3.

File: core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java (hunk 4 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java]: nit: this was lost, but I think is a good thing to have.

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java (hunk 10 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java]: opinions?

**Comment by bleskes** [comment]
LGTM. Thanks again @jasontedor

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java (hunk 10 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java]: I agree but would prefer that we do it in a follow up. That is, let's get the test running and stabilized and then enhance it. Are you okay with that @bleskes?

Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java]: OK!

File: core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java (hunk 4 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java]: I pushed 8b970d970ddfb32c27d51f436c4d6f7b35258540.

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java [hunk 13]
```diff
@@ -164,12 +164,17 @@ public synchronized void markAsProcessed(ClusterState state) {
                             currentMaster
                     );
                 }
-            } else if (state.supersedes(pendingState) && pendingContext.committed()) {
+            } else if (state.version() >= pendingState.version()) {
+                assert state.supersedes(pendingState) || (
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java]: is we're here, then the masters are equal. In this case supersedes == false, means version equality (because it's can't be lower. Also state.nodes().getMasterNodeId() is never null (and assigned to currentMaster.getId())) . I'm not sure what we're asserting than. Also note that it disables the next if clause, checking for uuid equality,  so maybe we want to do the next if clause first?

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java [hunk 14]
```diff
@@ -82,17 +82,22 @@
...
+    public PublishClusterStateAction(
+            Settings settings,
+            TransportService transportService,
+            Supplier<ClusterState> clusterStateSupplier,
+            NewPendingClusterStateListener listener,
+            DiscoverySettings discoverySettings,
+            ClusterName clusterName) {
         super(settings);
         this.transportService = transportService;
-        this.nodesProvider = nodesProvider;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: I think that node provider interface can go away (now that's were on java8 - function refs FTW)

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java [hunk 15]
```diff
@@ -395,11 +400,11 @@ void validateIncomingState(ClusterState incomingState, ClusterState lastSeenClus
             logger.warn("received cluster state from [{}] which is also master but with a different cluster name [{}]", incomingState.nodes().getMasterNode(), incomingClusterName);
             throw new IllegalStateException("received state from a node that is not part of the cluster");
         }
-        final DiscoveryNodes currentNodes = nodesProvider.nodes();
+        final DiscoveryNodes currentNodes = clusterStateSupplier.get().nodes();
 
         if (currentNodes.getLocalNode().equals(incomingState.nodes().getLocalNode()) == false) {
             logger.warn("received a cluster state from [{}] and not part of the cluster, should not happen", incomingState.nodes().getMasterNode());
-            throw new IllegalStateException("received state from a node that is not part of the cluster");
+            throw new IllegalStateException("received state from local node that does not match the current local node");
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: I think we want it read "received state *with* a local node that doesn't match the current one"

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java [hunk 16]
```diff
@@ -407,7 +412,7 @@ void validateIncomingState(ClusterState incomingState, ClusterState lastSeenClus
             final String message = String.format(
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: This is very close to the new check. I'm not sure we need it.

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java (hunk 14 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: I thought the same but can we keep it separate from this pull request?

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java [hunk 17]
```diff
@@ -416,6 +421,21 @@ void validateIncomingState(ClusterState incomingState, ClusterState lastSeenClus
             logger.warn(message);
             throw new IllegalStateException(message);
         }
+
+        final ClusterState state = clusterStateSupplier.get();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: I wonder if we should strengthen ZenDiscovery.shouldIgnoreOrRejectNewClusterState to check for version equality (and assert uuid) and use it [in line 410](https://github.com/elastic/elasticsearch/pull/17038/commits/cffc315dca66fa6af109a7cb5768acfb05e140f7#diff-594996d316ac34afe009d496db4abe34R410)  then all of this can be change to just throwing an exception if it doesn't like it.

File: core/src/test/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueueTests.java [hunk 18]
```diff
@@ -195,10 +195,11 @@ public void testQueueStats() {
                 highestCommitted = context.state;
             }
         }
+        assert highestCommitted != null;
 
         queue.markAsProcessed(highestCommitted);
-        assertThat(queue.stats().getTotal(), equalTo(states.size() - committedContexts.size()));
-        assertThat(queue.stats().getPending(), equalTo(states.size() - committedContexts.size()));
+        assertThat((long)queue.stats().getTotal(), equalTo(states.size() - (1 + highestCommitted.version())));
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueueTests.java]: why is this needed?

Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueueTests.java]: @bleskes Because uncommitted states with a lower version get cleaned now but they didn't before.

File: core/src/test/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateActionTests.java [hunk 19]
```diff
@@ -678,19 +695,27 @@ public void testOutOfOrderCommitMessages() throws Throwable {
...
-            if (channel.error.get() != null) {
-                throw channel.error.get();
+            if (largestVersionSeen < state.getVersion()) {
+                assertThat(channel.response.get(), equalTo((TransportResponse) TransportResponse.Empty.INSTANCE));
+                if (channel.error.get() != null) {
+                    throw channel.error.get();
+                }
+                largestVersionSeen = state.getVersion();
+            } else {
+                assertNotNull(channel.error.get());
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateActionTests.java]: can we add a comment as to why we expect an error?

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java (hunk 15 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: I pushed a3e007d697d069b5fd14efc1f95dbd125c145bc5.

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java (hunk 14 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: for **SURE**!

File: core/src/test/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueueTests.java (hunk 18 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueueTests.java]: doh. right.

File: core/src/test/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateActionTests.java (hunk 19 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateActionTests.java]: I pushed 3abf8177e87696102dbc9838b619a72328f32395.

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java (hunk 13 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java]: Removed assertion, reordered checks in c2ed5a1c9e4c7decb73eb026e7f7a8591909389f.

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java (hunk 17 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: I pushed 5fbba519fc1ad3acc88e42e68ae7615efe5b8c12.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: @bleskes I pushed 66cc2029cd4016afb6ac45587c5980d11431dd77.

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java (hunk 16 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: This check rejects incoming states from the same master that are out of order. I do not think that the new check covers this case. I think it's needed?

**Comment by jasontedor** [comment]
@bleskes I've addressed all comments from the latest feedback round.

File: core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java (hunk 16 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: With the new check (based on the ClusterService.state()), a cluster state with a lower/equal version than the current state will be rejected. If it has a higher version it will get in. If it is lower than the last seen cluster state it means that it will be cleaned when the first of these happen:
1) A commit message for this specific CS comes in.
2) A commit message for a higher CS comes in (including the lastSeenClusterState) and that CS is processed.

Is there a case where the lastSeenClusterState.supersedes based protection helps and isn't covered by the above?

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: > With the new check (based on the ClusterService.state()), a cluster state with a lower/equal version than the current state will be rejected. If it has a higher version it will get in.

Just to be clear, this is only if we are following a master. The line of code in question will reject older states from the same master that are out of order.

> If it is lower than the last seen cluster state it means that it will be cleaned when the first of these happen:
> 1) A commit message for this specific CS comes in.
> 2) A commit message for a higher CS comes in (including the lastSeenClusterState) and that CS is processed.
> Is there a case where the lastSeenClusterState.supersedes based protection helps and isn't covered by the above?

There is not.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: @bleskes I pushed 95feb40bef36ab5a21718062704c9039a17f29e0.

**Comment by bleskes** [comment]
LGTM again. Thanks for all the hard work @jasontedor
